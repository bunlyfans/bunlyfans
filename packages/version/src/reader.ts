import { Log } from "@bunlyfans/log/local";
import { PackageModel } from "./model";
import * as fs from "fs/promises";
import * as path from "path";

const log = Log({ scope: "Reader" });

export class Reader {
  constructor(private rootPath: string) {}

  async read(): Promise<[PackageModel, PackageModel[]]> {
    const root = await this.readFile(path.join(this.rootPath, "package.json"));

    const packages: PackageModel[] = [];

    if (root.workspaces) {
      for (const workspace of root.workspaces) {
        const packagesInWorkspace = await this.findInWorkspace(workspace);
        packages.push(...packagesInWorkspace);
      }
    }

    return [root, packages];
  }

  private async readFile(filePath: string): Promise<PackageModel> {
    try {
      const file = Bun.file(filePath);
      const data = await file.json();
      return new PackageModel(data.name, data.version, file, data.workspaces);
    } catch (error) {
      log.e(`Error while reading ${filePath}`);
      throw error;
    }
  }

  private async findInWorkspace(workspace: string): Promise<PackageModel[]> {
    try {
      const packagesPath = path.join(
        this.rootPath,
        workspace.replace(/\*/g, "")
      );

      // read the packages directory
      const packageDirs = await fs.readdir(packagesPath);

      const packages = [];

      for (const dir of packageDirs) {
        const packageJsonPath = path.join(packagesPath, dir, "package.json");
        packages.push(await this.readFile(packageJsonPath));
      }

      return packages;
    } catch (error) {
      log.e(`Error while reading ${workspace}`);
      throw error;
    }
  }
}
