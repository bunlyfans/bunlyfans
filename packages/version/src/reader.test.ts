import { describe, expect, test } from "bun:test";
import { Reader } from "./reader";

describe("reader", () => {
  test("read current repo", async () => {
    const rootVersion = new Reader("./src/mocks");
    const [root, packages] = await rootVersion.read();

    // to not local path to snapshot
    root.file = null as any;
    packages.forEach((i) => (i.file = null as any));

    expect(root).toMatchSnapshot();
    expect(packages).toMatchSnapshot();
  });
});
