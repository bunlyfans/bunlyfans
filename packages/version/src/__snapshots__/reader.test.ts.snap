// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`reader read current repo 1`] = `
PackageModel {
  "file": null,
  "name": "@mocks/mocks",
  "originalVersion": "0.1.0",
  "version": "0.1.0",
  "workspaces": [
    "packages/*",
  ],
}
`;

exports[`reader read current repo 2`] = `
[
  PackageModel {
    "file": null,
    "name": "@mocks/a",
    "originalVersion": "0.1.0",
    "version": "0.1.0",
    "workspaces": undefined,
  },
  PackageModel {
    "file": null,
    "name": "@mocks/b",
    "originalVersion": "0.2.0",
    "version": "0.2.0",
    "workspaces": undefined,
  },
]
`;
